File: frontend/types/transcript.ts
export interface TranscriptSummary {
  transcript_id: string;
  call_id: string;
  call_start_utc: string | null;
  call_end_utc: string | null;
  duration_seconds: number | null;
  call_outcome: string | null;
  vapi_end_reason: string | null;
  call_cost_usd: number | null;
  recording_url: string | null;
  call_summary: string | null;
  sentiment_overall: number | null;
  sentiment_score: number | null;
  transcript_raw: string | null;
  transcript_json: TranscriptMessage[] | Record<string, unknown> | null;
  tool_calls_log: unknown[] | null;
  dispatch_job_id: string | null;
  churn_risk_at_call_start: number | null;
  intervention_successful: boolean | null;
}

export interface TranscriptMessage {
  role?: string;
  speaker?: string;
  message?: string;
  text?: string;
  content?: string;
}

export type TranscriptDetail = TranscriptSummary;

export interface CustomerTranscriptsResponse {
  customer_id: string;
  transcripts: TranscriptSummary[];
}

File: backend/app/schemas/transcript.py
"""Call transcript API response schemas."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel

from app.models.call_transcript import CallTranscript


class TranscriptSummary(BaseModel):
    transcript_id: str
    call_id: str
    call_start_utc: str | None = None
    call_end_utc: str | None = None
    duration_seconds: int | None = None
    call_outcome: str | None = None
    vapi_end_reason: str | None = None
    call_cost_usd: float | None = None
    recording_url: str | None = None
    call_summary: str | None = None
    sentiment_overall: float | None = None
    sentiment_score: float | None = None
    transcript_raw: str | None = None
    transcript_json: list[Any] | dict[str, Any] | None = None
    tool_calls_log: list[Any] | None = None
    dispatch_job_id: str | None = None
    churn_risk_at_call_start: float | None = None
    intervention_successful: bool | None = None


class TranscriptDetail(TranscriptSummary):
    """Full transcript payload for a single call lookup."""


class CustomerTranscriptsResponse(BaseModel):
    customer_id: str
    transcripts: list[TranscriptSummary]


def transcript_to_summary(transcript: CallTranscript) -> TranscriptSummary:
    sentiment = (
        float(transcript.sentiment_overall)
        if transcript.sentiment_overall is not None
        else None
    )
    return TranscriptSummary(
        transcript_id=str(transcript.transcript_id),
        call_id=transcript.call_id,
        call_start_utc=transcript.call_start_utc.isoformat()
        if transcript.call_start_utc
        else None,
        call_end_utc=transcript.call_end_utc.isoformat()
        if transcript.call_end_utc
        else None,
        duration_seconds=transcript.duration_seconds,
        call_outcome=transcript.call_outcome,
        vapi_end_reason=transcript.vapi_end_reason,
        call_cost_usd=float(transcript.call_cost_usd)
        if transcript.call_cost_usd is not None
        else None,
        recording_url=transcript.recording_url,
        call_summary=transcript.call_summary,
        sentiment_overall=sentiment,
        sentiment_score=sentiment,
        transcript_raw=transcript.transcript_raw,
        transcript_json=transcript.transcript_json,
        tool_calls_log=transcript.tool_calls_log,
        dispatch_job_id=str(transcript.dispatch_job_id)
        if transcript.dispatch_job_id
        else None,
        churn_risk_at_call_start=float(transcript.churn_risk_at_call_start)
        if transcript.churn_risk_at_call_start is not None
        else None,
        intervention_successful=transcript.intervention_successful,
    )


def transcript_to_detail(transcript: CallTranscript) -> TranscriptDetail:
    return TranscriptDetail(**transcript_to_summary(transcript).model_dump())
